### ISSUE TITLE
[TSM] Panic during compaction

### ISSUE THREAD

**Issue by toddboom** [influxdb/influxdb#5257]
Doing a full-write load tests across two different servers, I encountered this panic:
```
panic: runtime error: slice bounds out of range

goroutine 7044023 [running]:
github.com/influxdb/influxdb/tsdb/engine/tsm1.(*indirectIndex).UnmarshalBinary(0xc84de30320, 0x7f7925d7b1fa, 0x166b6, 0x166be, 0x0, 0x0)
        /tmp/tmp.m6jFTgA2aM/src/github.com/influxdb/influxdb/tsdb/engine/tsm1/reader.go:622 +0x532
github.com/influxdb/influxdb/tsdb/engine/tsm1.(*mmapAccessor).init(0xc84fe81d60, 0x0, 0x0, 0x0, 0x0)
        /tmp/tmp.m6jFTgA2aM/src/github.com/influxdb/influxdb/tsdb/engine/tsm1/reader.go:858 +0x57c
github.com/influxdb/influxdb/tsdb/engine/tsm1.NewTSMReaderWithOptions(0x0, 0x0, 0xc8b7d9f040, 0x0, 0x0, 0x0)
        /tmp/tmp.m6jFTgA2aM/src/github.com/influxdb/influxdb/tsdb/engine/tsm1/reader.go:148 +0x2d8
github.com/influxdb/influxdb/tsdb/engine/tsm1.(*FileStore).Replace(0xc82f7440e0, 0xc829f0dd00, 0x2, 0x2, 0xc8c0cc9a10, 0x1, 0x1, 0x0, 0x0)
        /tmp/tmp.m6jFTgA2aM/src/github.com/influxdb/influxdb/tsdb/engine/tsm1/file_store.go:367 +0x41f
github.com/influxdb/influxdb/tsdb/engine/tsm1.(*DevEngine).compactTSMLevel.func1(0xc89597bfd0, 0xc820cd6f30, 0x1, 0x200010001, 0x0, 0xc829f0dd00, 0x2, 0x2)
        /tmp/tmp.m6jFTgA2aM/src/github.com/influxdb/influxdb/tsdb/engine/tsm1/engine.go:524 +0x79d
created by github.com/influxdb/influxdb/tsdb/engine/tsm1.(*DevEngine).compactTSMLevel
        /tmp/tmp.m6jFTgA2aM/src/github.com/influxdb/influxdb/tsdb/engine/tsm1/engine.go:535 +0x25f
```

Interestingly, it occurred on both boxes at nearly the same time - 4:47am on one and 4:52 on the other, which was ~16h into the test. Both tests were started at the same time. The first was 100k series using 5k batches, and 1m spread between timestamps (37GB written, 164 TSM files). The other was the same, but with 50k series (36GB written, 105 TSM files).

This test was performed with the following nightly build:
```
InfluxDB v0.10.0-nightly-6ccc416 (git: master 6ccc416ef0d3b0ba6d43328a2719886b66eb5e49, built 2016-01-01T05:00:47+0000)
```

**Comment by dswarbrick** [comment]
I haven't seen panics, but I'm experiencing reproducible OOM-kills almost exactly every 24h with the latest nightlies. I originally upgraded from 0.9.6.2 to resolve a bug where stale FDs were held open for deleted shards. The first nightly I tried (sometime pre Dec 23) resolved this. When I saw a bunch of commits relating to tsm1 from @jwilder on Dec 24, I figured I should update again, and that's when the problems started to occur. There is no panic recorded in the influxd.log, so these are the last few good lines in the log before influxd is OOM-killed and systemd restarts it:
```
[tsm1] 2016/01/02 20:07:15 beginning level 1 compaction of group 0, 2 TSM files
[tsm1] 2016/01/02 20:07:15 compacting level 1 group (0) /srv/influxdb/data/perfdb/default/53/000000651-000000001.tsm (#0)
[tsm1] 2016/01/02 20:07:15 compacting level 1 group (0) /srv/influxdb/data/perfdb/default/53/000000652-000000001.tsm (#1)
[tsm1] 2016/01/02 20:07:49 beginning full compaction of group 0, 11 TSM files
[tsm1] 2016/01/02 20:07:49 compacting full group (0) /srv/influxdb/data/perfdb/default/47/000000188-000000003.tsm (#0)
[tsm1] 2016/01/02 20:07:49 compacting full group (0) /srv/influxdb/data/perfdb/default/47/000000188-000000004.tsm (#1)
[tsm1] 2016/01/02 20:07:49 compacting full group (0) /srv/influxdb/data/perfdb/default/47/000000376-000000003.tsm (#2)
[tsm1] 2016/01/02 20:07:49 compacting full group (0) /srv/influxdb/data/perfdb/default/47/000000376-000000004.tsm (#3)
[tsm1] 2016/01/02 20:07:49 compacting full group (0) /srv/influxdb/data/perfdb/default/47/000000564-000000003.tsm (#4)
[tsm1] 2016/01/02 20:07:49 compacting full group (0) /srv/influxdb/data/perfdb/default/47/000000564-000000004.tsm (#5)
[tsm1] 2016/01/02 20:07:49 compacting full group (0) /srv/influxdb/data/perfdb/default/47/000000712-000000006.tsm (#6)
[tsm1] 2016/01/02 20:07:49 compacting full group (0) /srv/influxdb/data/perfdb/default/47/000000712-000000007.tsm (#7)
[tsm1] 2016/01/02 20:07:49 compacting full group (0) /srv/influxdb/data/perfdb/default/47/000000760-000000005.tsm (#8)
[tsm1] 2016/01/02 20:07:49 compacting full group (0) /srv/influxdb/data/perfdb/default/47/000000764-000000003.tsm (#9)
[tsm1] 2016/01/02 20:07:49 compacting full group (0) /srv/influxdb/data/perfdb/default/47/000000766-000000001.tsm (#10)
[monitor] 2016/01/02 20:11:02 failed to store statistics: timeout
[monitor] 2016/01/02 20:12:02 failed to store statistics: timeout
[tsm1] 2016/01/02 20:12:42 compacted level 1 group (0) into /srv/influxdb/data/perfdb/default/53/000000652-000000002.tsm.tmp (#0)
[tsm1] 2016/01/02 20:12:42 compacted level 1 group 0 of 2 files into 1 files in 5m26.251668631s
[tsm1] 2016/01/02 20:12:59 beginning level 1 compaction of group 0, 2 TSM files
[tsm1] 2016/01/02 20:12:59 compacting level 1 group (0) /srv/influxdb/data/perfdb/default/53/000000653-000000001.tsm (#0)
[tsm1] 2016/01/02 20:12:59 compacting level 1 group (0) /srv/influxdb/data/perfdb/default/53/000000654-000000001.tsm (#1)
[retention] 2016/01/02 20:13:49 retention policy enforcement check commencing
[retention] 2016/01/02 20:13:49 retention policy shard deletion check commencing
2016/01/02 20:13:50 Sending anonymous usage statistics to m.influxdb.com
Killed
2016/01/02 20:15:16 InfluxDB starting, version 0.10.0-nightly-6ccc416, branch master, commit 6ccc416ef0d3b0ba6d43328a2719886b66eb5e49, built 2016-01-01T05:00:47+0000
2016/01/02 20:15:16 Go version go1.5.2, GOMAXPROCS set to 16
2016/01/02 20:15:17 Using configuration at: /etc/influxdb/influxdb.conf
```
Also, this graph of various metrics on the influxdb host may prove useful. It shows two OOM-kills, indicated by the large spikes, then there is a dropout of about 30 mins in the network traffic while influxd restarts. Interesting to note is that the disk space utilization of the data partition increases significantly - which sorta smells like something weird happening during compaction.
![tsm1-oomkill](https://cloud.githubusercontent.com/assets/1667905/12077169/52efe334-b1d5-11e5-8d88-76d80022de57.png)

**Comment by toddboom** [comment]
@dswarbrick it might be a different issue, but can you tell me a bit more about the specs of the system you're running on? also, any information about the data you're feeding into InfluxDB (number of series, write throughput) would be helpful. thanks!

**Comment by dswarbrick** [comment]
@toddboom It's a kvm virtual machine with 32 GB RAM and 16 virtual cores (dedicated / pinned on the hypervisor host), Debian 8.x Jessie with kernel 4.3. As the graph shows, we are feeding about 8K points per second into it from about a total of about 1000 hosts, using the http line based protocol. We have about 30 measurements, and about 630K series, using a retention policy of 21 days. I don't have any CQs currently.

It has been running very smoothly with 0.9.6.2, apart from the stale FD issue when deleting shards as I mentioned, which resulted in the disk space not being freed. A pre-Dec-23 nightly of 0.10 also worked quite smoothly, and resolved said issue. It just seems that the last batch of commits by jwilder has unfortunately introduced a new issue.

**Comment by toddboom** [comment]
@dswarbrick the latest nightly build should have this fix - i'm curious to see if you still see the same behavior when testing with it or against `master`.

one other thing i wanted to check - since you're using a VM, i assume there's no swap space. is that correct?

**Comment by dswarbrick** [comment]
@toddboom I have just installed 0.10.0-nightly-d9ed54c. We shall see over the course of the next 24 hours if it behaves normally. The VM has 4GB swap space, which rarely gets touched during normal operation. To be honest, I was actually considering scaling the VM back from 32 GB down to 16 GB RAM, because most of it only seems to be used by vfs cache. On the other hand, RAM is cheap ;-).

**Comment by toddboom** [comment]
@dswarbrick great, definitely keep me posted. i did see an instance during some compactions where a lot of allocations caused a short-term memory spike and ultimately hit the swap on my test machine. this was with a higher cardinality data set, but it might still be related. we're looking at tuning out unnecessary allocations right now, so hopefully the nightlies will continue to improve over the next few days. thanks for all the feedback!

**Comment by jwilder** [comment]
@dswarbrick #5279 might help your OOM issue.

**Comment by dswarbrick** [comment]
@toddboom 0.10.0-nightly-d9ed54c was just oom-killed on my setup. I had installed 0.10.0-nightly-614a37c shortly prior, but not restarted it. The latter version is now starting up. We shall see in another 24h if the bug was fixed by #5279

**Comment by dswarbrick** [comment]
@toddboom @jwilder I am watching 0.10.0-nightly-614a37c eating RAM and CPU time as write this, and it looks like it is well on its way to being oom-killed shortly. The load started to increase substantially at 09:43 UTC, and looking in the log I can see some peculiar entries at that time:
```
[tsm1] 2016/01/07 09:41:29 compacting level 1 group (0) /srv/influxdb/data/perfdb/default/62/000000313-000000001.tsm (#0)
[tsm1] 2016/01/07 09:41:29 compacting level 1 group (0) /srv/influxdb/data/perfdb/default/62/000000314-000000001.tsm (#1)
[tsm1] 2016/01/07 09:43:33 compacting full group (0) /srv/influxdb/data/perfdb/default/21/000000189-000000004.tsm (#0)
[tsm1] 2016/01/07 09:43:33 compacting full group (0) /srv/influxdb/data/perfdb/default/21/000000189-000000005.tsm (#1)
[tsm1] 2016/01/07 09:43:33 compacting full group (0) /srv/influxdb/data/perfdb/default/21/000000378-000000004.tsm (#2)
[tsm1] 2016/01/07 09:43:33 compacting full group (0) /srv/influxdb/data/perfdb/default/21/000000378-000000005.tsm (#3)
...
[tsm1] 2016/01/07 09:45:21 compacting level 2 group (0) /srv/influxdb/data/perfdb/default/62/000000314-000000002.tsm (#0)
[tsm1] 2016/01/07 09:45:21 compacting level 2 group (0) /srv/influxdb/data/perfdb/default/62/000000316-000000002.tsm (#1)
[tsm1] 2016/01/07 09:46:36 compacting full group (0) /srv/influxdb/data/perfdb/default/25/000000189-000000004.tsm (#0)
[tsm1] 2016/01/07 09:46:36 compacting full group (0) /srv/influxdb/data/perfdb/default/25/000000189-000000005.tsm (#1)
[tsm1] 2016/01/07 09:46:36 compacting full group (0) /srv/influxdb/data/perfdb/default/25/000000377-000000004.tsm (#2)
...
[tsm1] 2016/01/07 09:58:39 compacting full group (0) /srv/influxdb/data/perfdb/default/41/000000752-000000004.tsm (#7)
[tsm1] 2016/01/07 09:58:39 compacting full group (0) /srv/influxdb/data/perfdb/default/41/000000780-000000002.tsm (#8)
[tsm1] 2016/01/07 10:00:06 compacting full group (0) /srv/influxdb/data/perfdb/default/43/000000189-000000003.tsm (#0)
[tsm1] 2016/01/07 10:00:06 compacting full group (0) /srv/influxdb/data/perfdb/default/43/000000189-000000004.tsm (#1)
```
etc. All of these full group compactions are being performed on *old shards*, which should have long since been compacted. I have a default retention policy of 21d, and as such these low-numbered shards are approaching 21 days old. The current shard number for my "perfdb" is 62.

The steadily increasing system load is accompanied by a huge increase in disk space usage. I think this is the best evidence I can provide so far, that there is something odd happening during some kind of 24h periodic DB maintenance.

**Comment by dswarbrick** [comment]
@jwilder I just saw influxd panic - no oom-kill today ;-)

Log file attached.
[influxd.txt](https://github.com/influxdata/influxdb/files/80978/influxd.txt)

**Comment by jwilder** [comment]
@dswarbrick This is helpful.  That panic is the same as the one in #5283.  I had a fix that I thought might have resolved that, but looks like it's not fixed.    

From your logs, it looks like there might be a bug in the full compaction planning and old shards are getting fully re-compacted after each reboot.  When a full compaction (or any compaction) runs, you'll see disk usage increase because we write new files from the files being compacted and then remove the old ones only after the new ones are fully written.  #5279 reduce the memory allocations that occurred while a compaction was running.  Since you didn't get an OOM, looks like that change helped, but exposed a new bug.

If your system is rebooting each night and that is causing full compactions to run for old shards after reboot, then you would definitely see a big memory spike and high disk usage since they could all be running concurrently.

I'm working on full compaction planning now so I'll take a look at this.
